lib/integrations/awsoidc/deploydbservice.go

```diff
@@ -0,0 +1,599 @@
+func upsertTask(ctx context.Context, clt DeployDBServiceClient, req DeployDBServiceRequest, configB64 string) (*ecsTypes.TaskDefinition, error) {
+	taskAgentContainerImage := fmt.Sprintf(teleportContainerImageFmt, req.TeleportVersion)
+
+	taskDefOut, err := clt.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+		Family: req.TaskName,
+		RequiresCompatibilities: []ecsTypes.Compatibility{
+			ecsTypes.CompatibilityFargate,
+		},
+		Cpu:    &taskCPU,
+		Memory: &taskMem,
+
+		NetworkMode:      ecsTypes.NetworkModeAwsvpc,
+		TaskRoleArn:      &req.TaskRoleARN,
+		ExecutionRoleArn: &req.TaskRoleARN,
+		ContainerDefinitions: []ecsTypes.ContainerDefinition{{
+			Command: []string{
+				"start",
+				"--config-string",
+				configB64,
+			},
+			EntryPoint: []string{"teleport"},
+			Image:      &taskAgentContainerImage,
+			Name:       &taskAgentContainerName,
+			LogConfiguration: &ecsTypes.LogConfiguration{
+				LogDriver: ecsTypes.LogDriverAwslogs,
+				Options: map[string]string{
+					"awslogs-group":         "ecs-" + *req.ClusterName,
+					"awslogs-region":        req.Region,
+					"awslogs-create-group":  "true",
+					"awslogs-stream-prefix": *req.ServiceName + "/" + *req.TaskName,
+				},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return taskDefOut.TaskDefinition, nil
+}
+
+// upsertCluster creates the cluster if it doesn't exist.
+// It will update the cluster if it doesn't have the required capacity provider (FARGATE)
+// It will re-create if its status is INACTIVE.
+// If the cluster status is not ACTIVE, an error is returned.
+// The cluster is returned.
+func upsertCluster(ctx context.Context, clt DeployDBServiceClient, clusterName string) (*ecsTypes.Cluster, error) {
+	var cluster *ecsTypes.Cluster
+
+	describeClustersResponse, err := clt.DescribeClusters(ctx, &ecs.DescribeClustersInput{
+		Clusters: []string{clusterName},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inactiveClusterExists := false
+	if len(describeClustersResponse.Clusters) > 0 {
+		cluster = &describeClustersResponse.Clusters[0]
+
+		// If the cluster was recently removed, it might still be listed but with "INACTIVE" status.
+		// Calling CreateCluster activates it.
+		// From AWS Docs:
+		// > INACTIVE The cluster has been deleted. Clusters with an INACTIVE status may remain discoverable in your account for a period of time.
+		if *cluster.Status == clusterStatusInactive {
+			inactiveClusterExists = true
+
+		} else if !slices.Contains(cluster.CapacityProviders, launcTypeFargateString) {
+			// Ensure the required capacity provider (Fargate) is available.
+			putClusterCPResp, err := clt.PutClusterCapacityProviders(ctx, &ecs.PutClusterCapacityProvidersInput{
+				Cluster:           &clusterName,
+				CapacityProviders: requiredCapacityProviders,
+				DefaultCapacityProviderStrategy: []ecsTypes.CapacityProviderStrategyItem{{
+					CapacityProvider: &launcTypeFargateString,
+				}},
+			})
```

Comment:
This seems risky to do to an existing cluster. If we depend on a specific capacity provider, isn't it reasonable to assume that whatever created the existing cluster might equally depend on its current configuration?

If I accidentally pointed teleport at an existing cluster which had incompatible configuration, I'd much prefer to have teleport generate an error about it rather than having teleport potentially silently break something.

Reply:
I changed the conditions to change the cluster.
It will only change if the cluster has the Teleport ownership Tags